### PR TITLE
Adding a retry policy for ReadTimeouts (not across all operations - but for reads and queries)

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**

### CODE REVIEW

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala [hunk 1]
```diff
@@ -33,13 +35,23 @@ import scala.reflect.ClassTag
 import scala.util.control.Breaks._
 import com.microsoft.azure.cosmosdb.rx.internal.NotFoundException
 
+private object CosmosDBConnection {
+  private val rnd = scala.util.Random
+
+  def getRandomRetryInterval(retryAttempt: Int): Int = {
+    rnd.nextInt(math.min(retryAttempt * 100, 5000))
```
Comment (moderakh) [inline src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:42]: why do we need a random retry?

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala [hunk 2]
```diff
@@ -33,13 +35,23 @@ import scala.reflect.ClassTag
...
 private[spark] case class CosmosDBConnection(config: Config) extends CosmosDBLoggingTrait with Serializable {
   private val maxPagesPerBatch =
     config.getOrElse[String](CosmosDBConfig.ChangeFeedMaxPagesPerBatch, CosmosDBConfig.DefaultChangeFeedMaxPagesPerBatch.toString).toInt
   private val clientConfig = ClientConfiguration(config)
 
   def getCollectionLink: String = {
-    executeWithRetryOnCollectionRecreate(() => CosmosDBConnectionCache.getOrReadContainerMetadata(clientConfig).selfLink)
+    executeWithRetryOnCollectionRecreate(
+      () => CosmosDBConnectionCache.getOrReadContainerMetadata(clientConfig).selfLink,
+      true)
```
Comment (ealsur) [inline src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:54]: nit: named parameters

**Review by moderakh — APPROVED** [review]
LGTM.

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala [hunk 3]
```diff
@@ -33,13 +35,23 @@ import scala.reflect.ClassTag
...
+}
+
 private[spark] case class CosmosDBConnection(config: Config) extends CosmosDBLoggingTrait with Serializable {
   private val maxPagesPerBatch =
     config.getOrElse[String](CosmosDBConfig.ChangeFeedMaxPagesPerBatch, CosmosDBConfig.DefaultChangeFeedMaxPagesPerBatch.toString).toInt
   private val clientConfig = ClientConfiguration(config)
 
   def getCollectionLink: String = {
-    executeWithRetryOnCollectionRecreate(() => CosmosDBConnectionCache.getOrReadContainerMetadata(clientConfig).selfLink)
+    executeWithRetryOnCollectionRecreate(
```
Comment (ealsur) [inline src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:52]: nit: do we have any tests that need to be updated?

**Review by ealsur — APPROVED** [review]

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:42]: Discussed offline - in the customer case ReadTimeout happened when getting PkRanges - adding random timeout to reduce chance that retries are happening at the same time

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:54]: Fixed

**Pull request merged**